dashboard/plugins/s3_usage/s3_stats.py

```diff
@@ -0,0 +1,193 @@
+
+        threading.Timer(next_run_after, self.run_refresher_task).start()
+
+    def start_refresher(self):
+        threading.Timer(1, self.run_refresher_task).start() 
+
+class S3StatsRefreshTask:
+    STATS_DB_PATH = '/var/run/s3_stats.db'
+    STATS_DB_TMP_PATH = '/var/run/s3_stats.db.tmp'
+    LOCK_PATH = '/var/run/s3_stats.db.lock'
+    CREATE_STMT = 'CREATE TABLE stats (path VARCHAR, files INTEGER, size_standard INTEGER, size_ia INTEGER, size_glacier INTEGER, depth INTEGER, is_leaf INTEGER)'
+    INSERT_STMT = 'INSERT INTO stats VALUES (?, ?, ?, ?, ?, ?, ?)'
+
+    def __init__(self, logger, buckets_regexp, aws_access_key_id, aws_secret_access_key, ttl):
+        s3 = boto3.client('s3', 
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key)
+        self.s3_aggregator = S3Aggregator(s3)
+        self.buckets = [bucket['Name'] for bucket in s3.list_buckets()['Buckets'] 
+                                            if re.match(buckets_regexp, bucket['Name'])]
+        self.logger = logger
+        self.ttl = ttl
+
+    def needs_refresh(self):
+        try:
+            return time.time() - os.stat(self.STATS_DB_PATH).st_mtime > self.ttl
+        except FileNotFoundError:
+            return True
+
+    def run(self):
+        with open(self.LOCK_PATH, 'w') as lock:
+            self.logger.debug("Trying to acquire the S3Stats lock")
+            fcntl.flock(lock, fcntl.LOCK_EX)
+            self.logger.debug("S3Stats Lock acquired")
+            if not self.needs_refresh():
+                self.logger.info("No need to refresh S3Stats, exiting")
+                return
+
+            self.sqlite = sqlite3.connect(self.STATS_DB_TMP_PATH, isolation_level="EXCLUSIVE")
+            self.sqlite.execute(self.CREATE_STMT)
+            self.sqlite.commit()
+            self.logger.info(f"Downloading stats for buckets {self.buckets}")
+
+            global_stats = KeyPrefix()
+            for bucket in self.buckets:
+                data = self.load_s3_bucket(bucket)
+                bucket_stats = self.dump_s3_bucket(data, bucket)
+                global_stats += bucket_stats
+
+            self.sqlite.execute(self.INSERT_STMT, ("", global_stats.files, global_stats.size_standard, global_stats.size_ia, global_stats.size_glacier, 0, 0))
+            self.sqlite.commit()
+            self.sqlite.close()
+            os.rename(self.STATS_DB_TMP_PATH, self.STATS_DB_PATH)
+
+    def insert_iterator(self, bucket, data):
+        for key in data:
+            key_fixed = key.rstrip('/')
+            is_leaf = sum([1 for path in data.keys() if path.startswith(key)]) == 1
+            yield f"/{bucket}{key_fixed}", data[key].files, \
+                data[key].size_standard, data[key].size_ia, data[key].size_glacier, \
+                len(key_fixed.split('/')), int(is_leaf)
+
+    def load_s3_bucket(self, bucket):
+        self.logger.info(f"Dumping bucket {bucket}")
+        start = time.time()
```

Comment:
In other plugins and in base we use specific Timer class, however, here we simply resolve to `time.time`. I would prefer to be more concise in that matter. A more generic TImer class would be the best option, however, we do not have it at the moment.

Reply:
Great idea, I will do it!